Recreate Backstage Pod on changes of ConfigMaps/Secrets configured in CR

gazarenkov opened the pull request:
## Description
This pull-request added following functionality:
- Adds label and annotation to ConfigMaps/Secrets configured in Backstage CR. 
- By default rhdh.redhat.com/ext-config-sync=true which allow controller watch ConfigMaps/Secrets 
- Enqueuing function calculates hash of the data, compares it with saved in the deployment's pod template annotation and calls reconcile in case they do not match (configuration changed)
- Since Pod template is changed k8s restarting it and reconciliation function saves new hash for the next cycle.

## Which issue(s) does this PR fix or relate to

https://github.com/janus-idp/operator/issues/236

## PR acceptance criteria

- [X] Tests
- [ ] Documentation TBD, scratched in the issue comments


## How to test changes / Special notes to the reviewer
- Create Backstage CR with configuration 
- Create ConfigMap(s)/Secret(s) referenced in CR
- Check if Operator added label rhdh.redhat.com/ext-config-sync=true and annotation rhdh.redhat.com/backstage-name=<CR name>
- Make sure Backstage instance is running
- Change anything in the data field of CMs/Secrets
- Make sure Backstage started recreating the Pod
- Make sure files/env vars with new content mounted to backstage container

rm3l commented on bundle/manifests/backstage-operator.clusterserviceversion.yaml:
```diff
@@ -21,7 +21,7 @@ metadata:
       ]
     capabilities: Seamless Upgrades
-    createdAt: "2024-04-01T18:15:06Z"
+    createdAt: "2024-05-02T09:16:44Z"
```
This is auto-generated when running `make bundle`.

gazarenkov replied:
well, it is a back side of storing generated files in git... I am not fan of it but that's what we decided to do for the time, can reconsider

gazarenkov commented on pkg/model/externalconfig.go:
```diff
@@ -0,0 +1,75 @@
+package model
+
+import (
+	"crypto/sha256"
```
that's what other do for exactly the same task, TBH I had no time to optimize it and do not think it will have big performance impact as it is called quite rarely, but I do not mind to replace it if you propose your option

coreydaley replied:
Huh, maybe sha256 is newer/more optimized now, carry on :)
```
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
md5             104574.53k   273858.63k   513204.40k   658448.10k   713978.01k   723875.91k
sha256          159849.35k   537076.76k  1377179.23k  2159443.74k  2600313.41k  2659953.40k
```

gazarenkov commented on pkg/model/runtime_test.go:
```diff
@@ -29,7 +31,14 @@ import (
+	for _, cfg := range runtimeConfig {
+		obj := cfg.ObjectFactory.newBackstageObject()
+		assert.NotNil(t, obj.EmptyObject())
+		// TODO
```
Yes, for the time EmptyObject() does not contain Kind, planning to add it in the future for tests and nicer logs

rm3l commented on controllers/spec_preprocessor.go:
```diff
@@ -69,47 +74,109 @@ func (r *BackstageReconciler) preprocessSpec(ctx context.Context, backstage bs.B
+	}
+
+	autoSync := true
+	autoSyncStr, ok := os.LookupEnv(AutoSyncEnvVar)
```
Also, could you document this (from a user standpoint, I mean)?

gazarenkov replied:
yes, on CSV for OLM (like custom images), directly on manifest otherwise (like fsGroup)

As commented, documentation will come later (with configuration docs)

rm3l commented on integration_tests/config-refresh_test.go:
```diff
@@ -0,0 +1,166 @@
+	It("refresh config", func() {
+
+		if !*testEnv.UseExistingCluster {
+			Skip("Skipped for not real cluster")
```
The test ended up failing after the timeout:

```
/home/asoro/work/projects/backstage/janus-idp/operator/integration_tests/config-refresh_test.go:56

  [FAILED] Timed out after 600.002s.

  The function passed to Eventually failed at /home/asoro/work/projects/backstage/janus-idp/operator/integration_tests/config-refresh_test.go:134 with:
  Expected
      <string>: app:
  to equal
      <string>: app:\n  backend:
  In [It] at: /home/asoro/work/projects/backstage/janus-idp/operator/integration_tests/config-refresh_test.go:139 @ 05/08/24 11:00:37.888
------------------------------
SSSS
------------------------------
[AfterSuite] 
/home/asoro/work/projects/backstage/janus-idp/operator/integration_tests/suite_test.go:128
  STEP: tearing down the test environment @ 05/08/24 11:00:37.892
[AfterSuite] PASSED [0.000 seconds]
------------------------------

Summarizing 1 Failure:
  [FAIL] when create backstage with external configuration [It] refresh config
  /home/asoro/work/projects/backstage/janus-idp/operator/integration_tests/config-refresh_test.go:139

Ran 1 of 11 Specs in 662.456 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 10 Skipped
--- FAIL: TestAPIs (662.46s)
FAIL

Ginkgo ran 1 suite in 11m5.056416574s

Test Suite Failed
make: *** [Makefile:137: integration-test] Error 1
```

Is it passing on your end?

gazarenkov replied:
Yes, it works

Ran 11 of 11 Specs in 532.223 seconds
SUCCESS! -- 11 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

However, real cluster related tests are flaky (same as on e2e tests), most of the time  (but not always) it is related to slow reconciliation. I am trying to reduce some moments but I would not like to increase timeout to infinity, so sometimes happened.

rm3l commented on bundle/manifests/backstage-operator.clusterserviceversion.yaml:
As commented out by the bot in https://github.com/janus-idp/operator/pull/351#issuecomment-2089985223, please make sure to review the RHDH CSV file in `.rhdh` and update it accordingly: https://github.com/gazarenkov/janus-idp-operator/blob/refresh_config/.rhdh/bundle/manifests/rhdh-operator.csv.yaml

gazarenkov replied:
ok

gazarenkov replied:
Could you be more specific please. What's the difference? I changed RBAC. Did I missed something?

rm3l replied:
I ran a manual diff to check this:

```diff
$ diff -U 1 <(yq '.spec.install.spec.clusterPermissions' bundle/manifests/backstage-operator.clusterserviceversion.yaml) <(yq '.spec.install.spec.clusterPermissions' .rhdh/bundle/manifests/rh
dh-operator.csv.yaml)

--- /proc/self/fd/11	2024-05-14 12:34:06.801140277 +0200
+++ /proc/self/fd/12	2024-05-14 12:34:06.802140285 +0200
@@ -5,24 +5,6 @@
         - configmaps
-        - services
-      verbs:
-        - create
-        - delete
-        - get
-        - list
-        - patch
-        - update
-        - watch
-    - apiGroups:
-        - ""
-      resources:
         - persistentvolumeclaims
         - persistentvolumes
-      verbs:
-        - get
-        - list
-        - watch
-    - apiGroups:
-        - ""
-      resources:
         - secrets
+        - services
       verbs:
@@ -32,3 +14,2 @@
         - list
-        - patch
         - update
@@ -44,3 +25,2 @@
         - list
-        - patch
         - update
@@ -56,3 +36,2 @@
         - list
-        - patch
         - update
@@ -89,3 +68,2 @@
         - routes
-        - routes/custom-host
       verbs:
@@ -95,3 +73,2 @@
         - list
-        - patch
         - update
@@ -110,2 +87,2 @@
         - create
-  serviceAccountName: backstage-controller-manager
+  serviceAccountName: rhdh-operator
```

The cluster permissions in `.rhdh` appear to be out of sync. But this looks like this issue was introduced before this PR. Okay, let's fix it in a subsequent issue/PR.

gazarenkov replied:
Yes, it is not the same, I can see it but it apparently can not be the same and as for RBAC it is definitely better than before :) , yea, let's back to it later.

rm3l replied:
I created a different issue to capture this, as I noticed another impact while testing this. https://github.com/janus-idp/operator/issues/360
I'm looking into this right now..

gazarenkov commented on controllers/backstage_controller.go:
```diff
@@ -53,12 +67,7 @@ type BackstageReconciler struct {
-	// Namespace allows to restrict the reconciliation to this particular namespace,
-	// and ignore requests from other namespaces.
-	// This is mostly useful for our tests, to overcome a limitation of EnvTest about namespace deletion.
-	Namespace string
```
Envtests use TestBackstageReconciler which contain namespace,  see https://github.com/janus-idp/operator/blob/c889b0e52bca79a7d5dfd5424165b735dc5103b7/integration_tests/suite_test.go#L69 so this workaround is not needed here anymore.

gazarenkov commented on integration_tests/config-refresh_test.go:
```diff
@@ -0,0 +1,166 @@
+		deleteNamespace(ctx, ns)
+	})
+
+	It("refresh config", func() {
```
This one would test if builder.WithPredicates works correct with LabelSelectorPredicate (does not call enqueue in this case). I would prefer to rely on K8s and avoid testing it's behaviour with costly tests (requiring real (i. e. not envtest) cluster and so quite long)

rm3l replied:
> This one would test if builder.WithPredicates works correct with LabelSelectorPredicate (does not call enqueue in this case). I would prefer to rely on K8s

Well, this is the case now, but how can I guarantee that this will continue to behave this way with future code changes? The thing I am mostly concerned about is: if I remove the code that handles the env var that controls the auto-sync mechanism, the expectations would no longer be correct, but the current tests would still pass. So we'd better catch this before the customer does.
Could this be covered with unit tests instead?

> avoid testing it's behaviour with costly tests (requiring real (i. e. not envtest) cluster and so quite long)

Also, I think that was the point of the nightly E2E tests.. It should be fine even if they take long (with a reasonable timeout of course), as long as they allow us to catch potential regressions earlier.

gazarenkov replied:
> Well, this is the case now, but how can I guarantee that this will continue to behave this way with future code changes? The thing I am mostly concerned about is: if I remove the code that handles the env var that controls the auto-sync mechanism, the expectations would no longer be correct, but the current tests would still pass. So we'd better catch this before the customer does.
> Could this be covered with unit tests instead?

I can not get your point, sorry.
Like I said I rely on K8s/go functionality, i.e. no label - no call, no annotation fitting backstage name - no call, no changes in configuration hash => no Pod recreating (deployment is identical), I do not test it here, the same as we do not usually test whether K8s's GET by name or Go's if work correct.  I am interested in testing functionality I added - i.e. when it all triggered - configuration changed (for other case we have other tests) ... It is a complex test and I definitely can not do it with unit or even envtest because it requires working deployment/pod controllers.
"if I remove the code that handles the env var that controls the auto-sync mechanism", FYI:
- env var does not control auto-sync mechanism, it is just opposite
- I guess, you can remove the code which handles re-creating env vars only with the piece of Pod controller  ;)

rm3l replied:
I understand, but the code changes introduce 2 paths: one (now default behavior) which is to always recreate the pod upon config changes, and another one (which is a valid and supported use case) is to not do it. Right now, only the former is tested.
Without going too deep into a real test, I was thinking of simply adding a test case to `cr-config_test.go`, to ensure that the `rhdh.redhat.com/ext-config-sync` label value can be `false` depending on the operator environment (like https://github.com/gazarenkov/janus-idp-operator/blob/refresh_config/integration_tests/cr-config_test.go#L187). Does that make sense?

gazarenkov replied:
Well, there is one place where this label is created/updated by operator (updating the CM/Secret):

````
		obj.GetLabels()[model.ExtConfigSyncLabel] = strconv.FormatBool(autoSync)
		obj.GetAnnotations()[model.BackstageNameAnnotation] = backstageName
		if err := r.Update(ctx, obj); err != nil {
````
Where autoSync is read from env variable or use 'true' by default.

```
        autoSync := true
	autoSyncStr, ok := os.LookupEnv(AutoSyncEnvVar)
	if ok {
		autoSync, _ = strconv.ParseBool(autoSyncStr)
	}

```
Did I understand you correct - you want me to make a test which 
- sets env var to 'false'
- reconsiles CR with ext CM/Secret 
- and check if CM/Secret's annotation is 'false' too?
(yes, I can do that if it helps)

rm3l replied:
Yes, that's exactly the path that I thought would make sense to cover: the condition that leads to setting this annotation to `false` (which should in turn prevent recreating the pod upon changes IIUC). If it makes sense of course :D
Thanks!

gazarenkov replied:
test added

gazarenkov commented on examples/refreshable-config.yaml:
```diff
@@ -0,0 +1,46 @@
+  labels:
+    rhdh.redhat.com/ext-config-sync: "true"
+  annotations:
+    rhdh.redhat.com/backstage-name: bs-rc-config
```
No

gazarenkov commented on controllers/spec_preprocessor.go:
```diff
@@ -69,47 +74,109 @@ func (r *BackstageReconciler) preprocessSpec(ctx context.Context, backstage bs.B
+		obj.GetLabels()[model.ExtConfigSyncLabel] != strconv.FormatBool(autoSync) {
+
+		obj.GetLabels()[model.ExtConfigSyncLabel] = strconv.FormatBool(autoSync)
+		obj.GetAnnotations()[model.BackstageNameAnnotation] = backstageName
```
I thought about that and I do not think creating many instances of Backstage and especially sharing configuration have any practical sense unless you have any good example why it is needed (I think we rather talk about one or two instances per cluster :) ).
So, I would rather document it as a limitation (which always can be workarounded with re-starting the pod) than add complexity.

rm3l replied:
Ok, makes sense. Thanks!

gazarenkov commented:
> Well, this is the case now, but how can I guarantee that this will continue to behave this way with future code changes? The thing I am mostly concerned about is: if I remove the code that handles the env var that controls the auto-sync mechanism, the expectations would no longer be correct, but the current tests would still pass. So we'd better catch this before the customer does.
> Could this be covered with unit tests instead?

I can not get your point, sorry.
Like I said I rely on K8s/go functionality, i.e. no label - no call, no annotation fitting backstage name - no call, no changes in configuration hash => no Pod recreating (deployment is identical), I do not test it here, the same as we do not usually test whether K8s's GET by name or Go's if work correct.  I am interested in testing functionality I added - i.e. when it all triggered - configuration changed (for other case we have other tests) ... It is a complex test and I definitely can not do it with unit or even envtest because it requires working deployment/pod controllers.
"if I remove the code that handles the env var that controls the auto-sync mechanism", FYI:
- env var does not control auto-sync mechanism, it is just opposite
- I guess, you can remove the code which handles re-creating env vars only with the piece of Pod controller  ;)  


> Is this resolved? I can still see some differences in the RBAC permissions between this CSV and the one under .rhdh..

Could you be more specific please. What's the difference? I changed RBAC. Did I missed something?

rm3l commented on .rhdh/bundle/manifests/rhdh-operator.csv.yaml:
@gazarenkov Could you rebase your branch onto `main`, to check if there is any diff regarding the permissions in the downstream CSV? I suspect there is a diff to handle here.. Thanks.

rm3l commented on controllers/preprocessor_test.go:
💯

rm3l reviewed (commented):
Thanks for the changes! Small comment about the permissions in the RHDH CSV..

rm3l commented on config/manager/default-config/app-config.yaml:
```diff
@@ -5,10 +5,10 @@ metadata:
-        connection:
-          password: ${POSTGRES_PASSWORD}
-          user: ${POSTGRES_USER}
+      #database:
```
nit: Let's remove this whole `backend.database` section if it is not needed..
